### DISCUSSION TITLE
Can't call numpy() on Tensor that requires grad - Value error

### DISCUSSION THREAD

**Discussion by Kullanicikukla61** [mrdbourke/pytorch-deep-learning#983] [Q&A]
My plot function givees that error: RuntimeError: Can't call numpy() on Tensor that requires grad. Use tensor.detach().numpy() instead.
I searched problem and find that: use that "y_pred = y_pred.detach().numpy()" before call function. This leads this: error: ValueError: x and y must be the same size. 
I searched discussions. However did not understand anything.
(I updated my code according to course materials but did not changed anyting)
(restart run all is the same.) 

    def plot_prediction(train_data = x_train,
                        train_labels= y_train,
                        test_data= x_test,
                        test_labels= y_test,
                        predictions= None):
      plt.figure(figsize=(10,7))
    
      plt.scatter(train_data,train_labels,c="b",s=4,label="Training label")
      plt.scatter(test_data,test_labels,c="g",s=4,label="Test label")
    
      if predictions is not None:
        plt.scatter(test_data, predictions, c="r", s=4, label="predictions")
    
      plt.legend(prop={"size":14} )
    
    
    
    torch.manual_seed(42)
    
    epochs = 1
    train_loss_values = []
    test_loss_values = []
    epoch_count = []
    
    
    for epoch in range(epochs):
      model0.train()
      y_pred= model0(x_train)        #forward pass
      loss = loss_fn(y_pred,y_train)      #calculate loss
      optimizer.zero_grad()             # optimeze to zero gradient
      loss.backward()                  #backpropogation according to parameters
      optimizer.step()                # step optimizer
      model0.eval()
      with torch.inference_mode():
        test_pred = model0(x_test)

        test_loss = loss_fn(test_pred,y_test.type(torch.float))
    
        if epoch % 10 == 0:
          epoch_count.append(epoch)
          train_loss_values.append(loss.detach().numpy())
          test_loss_values.append(test_loss.detach().numpy())
          print(f"Epoch: {epoch} | MAE Train Loss: {loss} | MAE Test Loss: {test_loss} ")
    
    
        y_pred = y_pred.detach().numpy()
        plot_prediction(predictions=y_pred)`
      

 there may be mistake in post but indentation is true. I am new at github.

**Comment by miladnorouziii** [comment]
try using this code. plt.plot(epochArray, torch.Tensor(training_losses).numpy(), label = "Train Loss")
                    plt.plot(epochArray, torch.Tensor(validation_losses).numpy(), label = "Validation Loss")
                    plt.title("Loss plots")
                    plt.xlabel("Epochs")
                    plt.ylabel("Loss")
                    plt.legend()
                    plt.show()
